ctru-rs/src/applets/swkbd.rs

```diff
@@ -675,6 +640,292 @@ impl SoftwareKeyboard {
+
+                let mut initial_text_cursor = swkbd_shared_mem_ptr.cast();
+
+                for code_point in utf16_iter {
+                    *initial_text_cursor = code_point;
+                    initial_text_cursor = initial_text_cursor.add(1);
+                }
+            }
+        }
+
+        if !extra.dict.is_null() {
+            swkbd.dict_offset = dict_off as _;
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    extra.dict,
+                    swkbd_shared_mem_ptr.add(dict_off).cast(),
+                    swkbd.dict_word_count as _,
+                )
+            };
+        }
+
+        if swkbd.initial_status_offset >= 0 {
+            swkbd.initial_status_offset = status_off as _;
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    extra.status_data,
+                    swkbd_shared_mem_ptr.add(status_off).cast(),
+                    1,
+                )
+            };
+        }
+
+        if swkbd.initial_learning_offset >= 0 {
+            swkbd.initial_learning_offset = learning_off as _;
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    extra.learning_data,
+                    swkbd_shared_mem_ptr.add(learning_off).cast(),
+                    1,
+                )
+            };
+        }
+
+        if extra.callback.is_some() {
+            swkbd.filter_flags |= SWKBD_FILTER_CALLBACK;
+        } else {
+            swkbd.filter_flags &= !SWKBD_FILTER_CALLBACK;
+        }
+
+        // Launch swkbd
+        unsafe {
+            swkbd.__bindgen_anon_1.reserved.fill(0);
+
+            if extra.callback.is_some() {
+                self.callback_data.extra = std::ptr::addr_of_mut!(extra);
+                self.callback_data.swkbd_shared_mem_ptr = swkbd_shared_mem_ptr;
+
+                aptSetMessageCallback(
+                    Some(Self::swkbd_message_callback),
+                    std::ptr::addr_of_mut!(self.callback_data).cast(),
+                );
+            }
+
+            aptLaunchLibraryApplet(
+                APPID_SOFTWARE_KEYBOARD,
+                swkbd as *mut _ as *mut _,
+                std::mem::size_of::<SwkbdState>(),
+                swkbd_shared_mem_handle,
+            );
+
+            if extra.callback.is_some() {
+                aptSetMessageCallback(None, std::ptr::null_mut());
+            }
+
+            let _ = svcCloseHandle(swkbd_shared_mem_handle);
+        }
+
+        let button = match swkbd.result {
+            SWKBD_D1_CLICK0 | SWKBD_D2_CLICK0 => SWKBD_BUTTON_LEFT,
+            SWKBD_D2_CLICK1 => SWKBD_BUTTON_MIDDLE,
+            SWKBD_D0_CLICK | SWKBD_D1_CLICK1 | SWKBD_D2_CLICK2 => SWKBD_BUTTON_RIGHT,
+            _ => SWKBD_BUTTON_NONE,
+        };
+
+        if swkbd.text_length > 0 {
+            let text16 = unsafe {
+                widestring::Utf16Str::from_slice_unchecked(std::slice::from_raw_parts(
```

Comment:
I assume the `_unchecked` here means that we are trusting either the OS or `libctru` to give us a valid string, right? Just wondering if an extra check here might be worth a little bit of safety

Reply:
Yeah, we're trusting the Software Keyboard to give us proper UTF-16 text here and I don't think there's any reason to doubt it would fail at that job.